README.md
<img src="https://opentable.github.io/spur/logos/Spur-Errors.png?rand=1" width="100%" alt="Spur: Errors" />

Common error builder utility for [Node.js](http://nodejs.org/). Contains common error types, and stack trace tracking to support more detailed error messages.

[![npm version](https://badge.fury.io/js/spur-errors.svg)](http://badge.fury.io/js/spur-errors)
[![Build Status](https://travis-ci.org/opentable/spur-errors.svg?branch=master)](https://travis-ci.org/opentable/spur-errors)
[![Dependencies](https://david-dm.org/opentable/spur-errors.svg)](https://david-dm.org/opentable/spur-errors)
[![devDependency Status](https://david-dm.org/opentable/spur-errors/dev-status.svg)](https://david-dm.org/opentable/spur-errors?type=dev)

# About the Spur Framework

The Spur Framework is a collection of commonly used Node.JS libraries used to create common application types with shared libraries.

[Visit NPMJS.org for a full list of Spur Framework libraries](https://www.npmjs.com/browse/keyword/spur-framework) >>

# Usage

Supports active Node versions in the [LTS Schedule](https://github.com/nodejs/LTS#lts-schedule). ([view current versions](.travis.yml))

## Install from NPM

```shell
$ npm install --save spur-errors
```

## Require and use the module

```javascript
let SpurErrors = require("spur-errors");

SpurErrors.NotFoundError.create("could not find it");
```
# API

The API is designed to be daisy chained with all of the following base commands that are a part of all of the error types.

## Base Object Commands

#### .create(message, nativeError) -> instance

Creates an instance of a SpurError for the type used.

```javascript
try {
  ...
}
catch(err) {
  SpurErrors.NotFound.create("Some error", err);
}
```

#### .setErrorCode(errorCode) -> instance

Sets an error code to later be used by error handlers.

```javascript
SpurErrors.NotFound.create("Not found").setErrorCode("leaf_error");
```

#### .setMessage(message) -> instance

Overrides the error message passed in.

```javascript
SpurErrors.NotFound.create("Not found").setMessage("Unable to find the restaurant.");
```

#### .setStatusCode(statusCode) -> instance

Setting the response status code to be sent back down to the client.

```javascript
SpurErrors.NotFound.create("Not found").setStatusCode(404);
```

#### .setData(data) -> instance

Sets customizable data that can be used down the error stack chain.

```javascript
SpurErrors.NotFound.create("Not found").setData({headers: req.headers});
```

## Properties

| Property      | Description                                                                         |
| :------------ | :---------------------------------------------------------------------------------- |
| internalError | The original error object passed in                                                 |
| message       | Either passed in during the create call or during the parsing of the internal error |
| stack         | Parsed from the originally passed in internal error                                 |
| errorCode     | Custom error code                                                                   |
| statusCode    | Custom status code to be used by the Express.JS response                            |
| data          | Custom data object to be used anyone in the flow                                    |

## Error Types

| Error Type              | Status Code | Message                   | Error Code                |
| :---------------------- | :---------- | :------------------------ | :------------------------ |
| ValidationError         | 400         | Validation Error          | validation_error          |
| UnauthorizedError       | 401         | Unauthorized Error        | unauthorized_error        |
| ForbiddenError          | 403         | Forbidden Error           | forbidden_error           |
| NotFoundError           | 404         | Not Found Error           | not_found_error           |
| MethodNotAllowedError   | 405         | Method not allowed        | method_not_allowed_error  |
| RequestTimeoutError     | 408         | Request Timeout Error     | request_timeout_error     |
| AlreadyExistsError      | 409         | Already Exists Error      | already_exists_error      |
| InternalServerError     | 500         | Internal Server Error     | internal_server_error     |
| BadGatewayError         | 502         | Bad Gateway Error         | bad_gateway_error         |
| ServiceUnavailableError | 503         | Service Unavailable Error | service_unavailable_error |

### Error type example

```javascript
SpurErrors.ValidationError.create("Invalid input");
// => {statusCode: 400, message: "Validation Error", errorCode: "validation_error", ....}
```

# Maintainers

This library is maintained by

  - Agustin Colchado – ***[@acolchado](https://github.com/acolchado)***

## Collaborators

- Ash – ***[@ssetem](https://github.com/ssetem)***
- Timmy Willison – ***[@timmywil](https://github.com/timmywil)***


# Contributing

## We accept pull requests

Please send in pull requests and they will be reviewed in a timely manner. Please review this [generic guide to submitting a good pull requests](https://github.com/blog/1943-how-to-write-the-perfect-pull-request). The only things we ask in addition are the following:

 * Please submit small pull requests
 * Provide a good description of the changes
 * Code changes must include tests
 * Be nice to each other in comments. :innocent:

## Style guide

The majority of the settings are controlled using an [EditorConfig](.editorconfig) configuration file. To use it [please download a plugin](http://editorconfig.org/#download) for your editor of choice.

## All tests should pass

To run the test suite, first install the dependancies, then run `npm test`

```bash
$ npm install
$ npm test
```

# License

[MIT](LICENSE)


CHANGELOG.md
[Changes (changelog) to this module are tracked in the releases](https://github.com/opentable/spur-errors/releases) >>


test/globals.js
const path = require('path');
const chai = require('chai');

global.chai = chai;
global.expect = chai.expect;
global.srcDir = path.resolve(__dirname, '../src');
global.rootDir = path.resolve(__dirname, '../');

process.env.NODE_ENV = 'test';

process.setMaxListeners(1000);


test/fixtures/Callee.js
const SpurErrors = require('../../src/SpurErrors');

module.exports = {

  run() {
    return SpurErrors.NotFoundError.create();
  }
};


test/unit/ModuleSpec.js
const SpurErrorsModule = require('../../');
const SpurErrorsSource = require('../../src/SpurErrors');

describe('Module Integration', () => {
  it('the module should be referenced from the root using what\'s defined in the package', () => {
    expect(SpurErrorsModule).to.exist;
  });

  it('the module should be the same as the expected source', () => {
    expect(SpurErrorsModule).to.equal(SpurErrorsSource);
  });
});


src/BaseError.js
const BaseError = {
  create(message, internalError) {
    Error.captureStackTrace(this);
    const error = Object.assign(new Error(), BaseError);
    if (message) {
      error.message = message;
    }
    if (internalError) {
      error.internalError = internalError;
    }
    return error;
  },

  extend(statusCode, defaultMessage, errorCode) {
    return {
      statusCode,
      create(message, internalError) {
        return Object.assign(
          BaseError.create(message || defaultMessage, internalError),
          { statusCode, errorCode }
        );
      }
    };
  },

  set internalError(internalError) {
    if (internalError && internalError.stack) {
      this.stack += '========================================================================\n\n';
      this.stack += internalError.stack;
    }
  },

  setErrorCode(errorCode) {
    this.errorCode = errorCode;
    return this;
  },

  setMessage(message) {
    this.message = message;
    return this;
  },

  setStatusCode(statusCode) {
    this.statusCode = statusCode;
    return this;
  },

  setData(data) {
    this.data = data;
    return this;
  }
};

module.exports = BaseError;


src/SpurErrors.js
const BaseError = require('./BaseError');

const SpurErrors = {
  BaseError,

  ValidationError: BaseError.extend(400, 'Validation Error', 'validation_error'),
  UnauthorizedError: BaseError.extend(401, 'Unauthorized Error', 'unauthorized_error'),
  ForbiddenError: BaseError.extend(403, 'Forbidden Error', 'forbidden_error'),
  NotFoundError: BaseError.extend(404, 'Not Found Error', 'not_found_error'),
  MethodNotAllowedError: BaseError.extend(405, 'Method Not Allowed', 'method_not_allowed_error'),
  RequestTimeoutError: BaseError.extend(408, 'Request Timeout Error', 'request_timeout_error'),
  AlreadyExistsError: BaseError.extend(409, 'Already Exists Error', 'already_exists_error'),
  InternalServerError: BaseError.extend(500, 'Internal Server Error', 'internal_server_error'),
  BadGatewayError: BaseError.extend(502, 'Bad Gateway Error', 'bad_gateway_error'),
  ServiceUnavailableError: BaseError.extend(503, 'Service Unavailable Error', 'service_unavailable_error'),

  errorByStatusCode(statusCode) {
    const errorNames = Object.keys(SpurErrors);
    let spurError;
    errorNames.forEach((name) => {
      const error = SpurErrors[name];
      if (error.statusCode === statusCode) {
        spurError = error;
      }
    });
    return spurError;
  }
};

module.exports = SpurErrors;
